sql/expression/function/load_file.go

```diff
@@ -0,0 +1,188 @@
+)
+
+type LoadFile struct {
+	fileName sql.Expression
+}
+
+var _ sql.FunctionExpression = (*LoadFile)(nil)
+
+func NewLoadFile(ctx *sql.Context, fileName sql.Expression) sql.Expression {
+	return &LoadFile{
+		fileName: fileName,
+	}
+}
+
+func (l LoadFile) Resolved() bool {
+	return true
+}
+
+func (l LoadFile) String() string {
+	return fmt.Sprintf("LOAD_FILE(%s)", l.fileName)
+}
+
+func (l LoadFile) Type() sql.Type {
+	return sql.LongBlob
+}
+
+func (l LoadFile) IsNullable() bool {
+	return false
+}
+
+// TODO: Allow FILE privileges for GRANT
+func (l LoadFile) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	dir, err := getSecureFilePriv(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Read the file: Ensure it fits the max byte size
+	// According to the mysql spec we must return NULL if the file is too big.
+	file, err := l.getFile(ctx, row, dir)
+
+	if err != nil {
+		return nil, handleFileErrors(err)
+	}
+	if file == nil {
+		return nil, nil
+	}
+
+	defer file.Close()
+
+	size, isTooBig, err := isFileTooBig(ctx, file)
+	if err != nil {
+		return nil, err
+	}
+	if isTooBig {
+		return nil, nil
+	}
+
+	// Finally, read the file
+	data := make([]byte, size)
+	_, err = file.Read(data)
+	if err != nil {
+		return nil, err
+	}
+
+	return data, nil
+}
+
+func (l *LoadFile) getFile(ctx *sql.Context, row sql.Row, secureFileDir string) (*os.File, error) {
+	fileName, err := l.fileName.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	// If the secure_file_priv directory is not set, just read the file from whatever directory it is in
+	// Otherwise determine whether the file is in the secure_file_priv directory.
+	if secureFileDir == "" {
+		return os.Open(fileName.(string))
+	}
+
+	// Open the two directories (secure_file_priv and the file dir) and validate they are the same.
+	sDir, err := os.Open(secureFileDir)
+	if err != nil {
+		return nil, err
+	}
+
+	sStat, err := sDir.Stat()
+	if err != nil {
+		return nil, err
+	}
+
+	ffDir, err := os.Open(filepath.Dir(fileName.(string)))
+	if err != nil {
+		return nil, err
+	}
+
+	fStat, err := ffDir.Stat()
+	if err != nil {
+		return nil, err
+	}
+
+	// If the two directories are not equivalent we return nil
+	if !os.SameFile(sStat, fStat) {
+		return nil, nil
+	}
+
+	return os.Open(fileName.(string))
+}
+
+func getSecureFilePriv(ctx *sql.Context) (string, error) {
+	val, err := ctx.Session.GetSessionVariable(ctx, "secure_file_priv")
+	if err != nil {
+		return "", err
+	}
+
+	return val.(string), nil
+}
+
+func isFileTooBig(ctx *sql.Context, file *os.File) (int64, bool, error) {
+	fi, err := file.Stat()
+	if err != nil {
+		return -1, false, err
+	}
+
+	maxByteSize, err := getMaxByteSize(ctx)
+	if err != nil {
+		return -1, false, err
+	}
+
+	return fi.Size(), fi.Size() > maxByteSize, nil
+}
+
+func getMaxByteSize(ctx *sql.Context) (int64, error) {
+	val, err := ctx.Session.GetSessionVariable(ctx, "max_allowed_packet")
+
+	if err != nil {
+		return 0, err
+	}
```

Comment:
These "_getSessionVar_" functions are actually worse than just using their contents directly. You _generally_ want to pull code out into its own function for one of the following reasons:
* The code will/might be used in multiple places
* The calling function is long
* You're breaking up a function into logical parts
* The code is unclear/non-trivial, so you want space for additional comments and/or a descriptive function name explains the contained code

Of course there are many, many more reasons for/against pulling out code, but these are some common ones. And these functions actually go against the ones that I've listed:
* This are only used once and won't be used anywhere else
* Size of the calling function is fine
* Nothing to break up as you're literally trading one function call for another (also causing two consecutive error checks with no benefit)
* `GetSessionVariable(ctx, "secure_file_priv")` is actually more descriptive than `getSecureFilePriv`, as it's clear to see that we're pulling it from the context (along with the exact variable name)

Just delete these functions and replace them with the contained code.

Reply:
Gotcha makes sense. Thanks for detailed comment here